include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+            if (cursor + wipeSize < header.m_commandAllocator.get_total_size())
+                memset(m_data + cursor + offsetof(IGPUCommandPool::ICommand, m_size), 0, sizeof(IGPUCommandPool::ICommand::m_size));
+        }
+    };
+    static_assert(sizeof(CCommandSegment) == COMMAND_SEGMENT_SIZE);
+
+    class CBeginCmd;
+    class CBindIndexBufferCmd;
+    class CDrawIndirectCmd;
+    class CDrawIndexedIndirectCmd;
+    class CDrawIndirectCountCmd;
+    class CDrawIndexedIndirectCountCmd;
+    class CBeginRenderPassCmd;
+    class CPipelineBarrierCmd;
+    class CBindDescriptorSetsCmd;
+    class CBindComputePipelineCmd;
+    class CUpdateBufferCmd;
+    class CResetQueryPoolCmd;
+    class CWriteTimestampCmd;
+    class CBeginQueryCmd;
+    class CEndQueryCmd;
+    class CCopyQueryPoolResultsCmd;
+    class CBindGraphicsPipelineCmd;
+    class CPushConstantsCmd;
+    class CBindVertexBuffersCmd;
+    class CCopyBufferCmd;
+    class CCopyBufferToImageCmd;
+    class CBlitImageCmd;
+    class CCopyImageToBufferCmd;
+    class CExecuteCommandsCmd;
+    class CDispatchIndirectCmd;
+    class CWaitEventsCmd;
+    class CCopyImageCmd;
+    class CResolveImageCmd;
+    class CClearColorImageCmd;
+    class CClearDepthStencilImageCmd;
+    class CFillBufferCmd;
+    class CSetEventCmd;
+    class CResetEventCmd;
+    class CWriteAccelerationStructurePropertiesCmd;
+    class CBuildAccelerationStructuresCmd; // for both vkCmdBuildAccelerationStructuresKHR and vkCmdBuildAccelerationStructuresIndirectKHR
+    class CCopyAccelerationStructureCmd;
+    class CCopyAccelerationStructureToOrFromMemoryCmd; // for both vkCmdCopyAccelerationStructureToMemoryKHR and vkCmdCopyMemoryToAccelerationStructureKHR
+
+    IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx)
+        : IBackendObject(std::move(dev)), m_commandSegmentPool(COMMAND_SEGMENTS_PER_BLOCK* COMMAND_SEGMENT_SIZE, 0u, MAX_COMMAND_SEGMENT_BLOCK_COUNT, MIN_POOL_ALLOC_SIZE),
+        m_flags(_flags), m_familyIx(_familyIx)
+    {}
+
+    inline core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
+    inline uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+
+    // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
+    // Vulkan: const VkCommandPool*
+    virtual const void* getNativeHandle() const = 0;
+
+    template <typename Cmd, typename... Args>
+    Cmd* emplace(CCommandSegment::Iterator& segmentListHeadItr, CCommandSegment*& segmentListTail, Args&&... args)
+    {
+        if (segmentListTail == nullptr)
+        {
+            void* cmdSegmentMem = m_commandSegmentPool.allocate(COMMAND_SEGMENT_SIZE, alignof(CCommandSegment));
+            if (!cmdSegmentMem)
+            {
+                assert(false);
+                return nullptr;
+            }
+
+            segmentListTail = new (cmdSegmentMem) CCommandSegment;
```

Comment:
is there a reason to not use
```cpp
segmentListTail = m_commandSegmentPool.emplace<CCommandSegment>();
if (!segmentListTail)
{
   assert(false);
   segmentListHeadItr = {};
   return nullptr;
}
segmentListHeadItr = segmentListTail->begin();
```

Reply:
Related: https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r1043992459

Introducing the `newSegment` lambda could be fine

Reply:
> is there a reason to not use emplace instead of allocate?

I think there isn't, as long as guarantees the alignment we want.

Reply:
I think the `newSegment` lambda has a mistake here:
```C++
if (segmentListTail)
         segmentListTail->setNext(next);
segmentListTail = next;
segmentListHeadItr = next->begin();
```
because we don't want to update the head of the list to the newly allocated segment, every time we make a new segment.

So I changed it a bit, mainly putting the above in an `else` block and changing the return type to `CCommandSegment*`.